File: scripts/reporting_guardrails.py
"""
reporting_guardrails.py
Verified Outcome Reporting Guardrails for SEPT.
"""

from dataclasses import dataclass
from typing import Optional, Literal

DeliveryStatus = Literal["queued", "attempting_to_send", "sent", "failed_to_send"]

@dataclass
class MessageDeliveryRecord:
    message_id: str
    recipient: str
    status: DeliveryStatus
    gateway_confirmed: bool = False
    error_message: Optional[str] = None

class VerifiedReportingEngine:
    @staticmethod
    def get_delivery_status(record: MessageDeliveryRecord) -> str:
        if record.gateway_confirmed and record.status == "sent":
            return f"Message {record.message_id} to {record.recipient}: Sent (Delivery Confirmed)"
        if record.status == "failed_to_send":
            return f"Message {record.message_id} to {record.recipient}: failed to send ({record.error_message or 'gateway dispatch error'})"
        if record.status == "queued":
            return f"Message {record.message_id} to {record.recipient}: queued"
        return f"Message {record.message_id} to {record.recipient}: attempting to send"

    @staticmethod
    def report_search(search_executed: bool, results_count: int, query: str) -> str:
        if not search_executed:
            return f'Search has not been run for query "{query}".'
        if results_count == 0:
            return f'Search executed: No listing found for "{query}".'
        return f'Search executed: Found {results_count} listing(s) for "{query}".'

    @staticmethod
    def transcribe_voice_note(audio_processed: bool, transcript: Optional[str] = None) -> str:
        if not audio_processed or not transcript or not transcript.strip():
            return "I couldn't process this voice note."
        return transcript.strip()

    @staticmethod
    def handle_voice_pushback() -> str:
        return "I couldn't process this voice note."

File: src/domain/reportingGuardrails.ts
/**
 * ReportingGuardrails — Enforces verified reporting invariants for SEPT.
 *
 * Invariants:
 *  1. Never claim "Sent" without verified delivery confirmation from the gateway.
 *     State must be "queued" or "attempting to send" while in transit, and "failed to send" on error.
 *  2. "No listing found" requires a genuine search execution against the catalog/network.
 *     If no search was executed, explicitly report that search has not been run.
 *  3. Voice note audio failure returns exact invariant: "I couldn't process this voice note."
 *     Strictly refuse speculation or hallucination under user pushback.
 */

export type DeliveryState = "queued" | "attempting_to_send" | "sent" | "failed_to_send";

export interface OutboundMessageState {
  messageId: string;
  recipientJid: string;
  state: DeliveryState;
  gatewayAckReceived: boolean;
  failureReason?: string;
}

export class VerifiedReportingEngine {
  /**
   * Determine truthful message delivery status string.
   */
  static getDeliveryStatusReport(msg: OutboundMessageState): string {
    if (msg.gatewayAckReceived && msg.state === "sent") {
      return `Message ${msg.messageId} to ${msg.recipientJid}: Sent (Delivery Confirmed)`;
    }
    if (msg.state === "failed_to_send") {
      return `Message ${msg.messageId} to ${msg.recipientJid}: failed to send (${msg.failureReason || "gateway error"})`;
    }
    if (msg.state === "queued") {
      return `Message ${msg.messageId} to ${msg.recipientJid}: queued`;
    }
    return `Message ${msg.messageId} to ${msg.recipientJid}: attempting to send`;
  }

  /**
   * Validate search execution before reporting outcome.
   */
  static reportSearchResult(searchExecuted: boolean, resultsCount: number, query: string): string {
    if (!searchExecuted) {
      return `Search has not been run for query "${query}".`;
    }
    if (resultsCount === 0) {
      return `Search executed: No listing found for "${query}".`;
    }
    return `Search executed: Found ${resultsCount} listing(s) for "${query}".`;
  }

  /**
   * Handle voice note processing outcome.
   */
  static handleVoiceNoteTranscription(audioProcessed: boolean, transcript?: string): string {
    if (!audioProcessed || !transcript || transcript.trim().length === 0) {
      return "I couldn't process this voice note.";
    }
    return transcript;
  }

  /**
   * Strict refusal against user pushback requesting speculation on unprocessable audio.
   */
  static handlePushbackOnAudio(): string {
    return "I couldn't process this voice note.";
  }
}
